working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1726,6 +1801,61 @@ behavior.
     [null safety][]. In fully migrated programs, exhaustiveness checking is
     sound and it isn't possible to reach this runtime error.*
 
+#### Pattern-for statement
+
+A statement of the form:
+
+```dart
+for (<patternDeclaration>; <condition>; <increment>) <statement>
+```
+
+Is executed similar to a traditional for loop except that multiple variables may
+be declared by the pattern instead of just one. As with a normal for loop, those
+variables are freshly bound to new values at each iteration so that if a
+function closes over a variable, it captures the value at the current iteration
+and is not affected by later iteration.
+
+The increment clause is evaluated in a scope where all variables declared in the
+pattern are freshly bound to new variables holding the current iteration's
+values. If the increment clause assigns to any of the variables declared by the
+pattern, those become the values bound to those variables in the next iteration.
+For example:
+
+```dart
+var fns = <Function()>[];
+for (var (a, b) = (0, 1); a <= 13; (a, b) = (b, a + b)) {
+  fns.add(() {
+    print(a);
+  });
+}
+
+for (var fn in fns) {
+  fn();
+}
+```
+
+This prints `0`, `1`, `1`, `2`, `3`, `5`, `8`, `13`.
+
+#### Pattern-for-in statement
+
+A statement of the form:
+
+```dart
+for (<keyword> <pattern> in <expression>) <statement>
+```
+
+Where `<keyword>` is `var` or `final` is interpreted as the following code,
+where `id1` and `id2` are fresh identifiers:
+
+```
+var id1 = <expression>;
```

Comment:
Current spec says:

T id1 = e;
var id2 = id1.iterator;
while (id2.moveNext()) {
D id = id2.current;
{ S }
}
Dart Programming Language Specification 192
If the static type of e is a top type (15.2) then T is Iterable<dynamic>, otherwise T is the static type of e. It is a compile-time error if T is not assignable
to Iterable<dynamic>.

We probably want to keep the constraints around Iterable?